File: src/main/java/com/vandendaelen/nicephore/clipboard/impl/WindowsClipboardManagerImpl.java
package com.vandendaelen.nicephore.clipboard.impl;

import com.profesorfalken.jpowershell.PowerShell;
import com.vandendaelen.nicephore.clipboard.ClipboardManager;

import java.io.File;

public class WindowsClipboardManagerImpl implements ClipboardManager {
    private static WindowsClipboardManagerImpl INSTANCE;
    private final PowerShell session;

    public WindowsClipboardManagerImpl() {
        this.session = PowerShell.openSession();
    }

    @Override
    public void clipboardImage(File screenshot) {
        final String command = "[Reflection.Assembly]::LoadWithPartialName('System.Drawing');\n" +
                "[Reflection.Assembly]::LoadWithPartialName('System.Windows.Forms');\n" +
                "\n" +
                String.format("$filename = \"%s\";\n", screenshot.getAbsolutePath()) +
                "$file = get-item($filename);\n" +
                "$img = [System.Drawing.Image]::Fromfile($file);\n" +
                "[System.Windows.Forms.Clipboard]::SetImage($img);";

        final Thread clipboardThread = new Thread(() -> session.executeCommand(command));
        clipboardThread.start();
    }

    public static WindowsClipboardManagerImpl getInstance() {
        if (INSTANCE == null) {
            INSTANCE = new WindowsClipboardManagerImpl();
        }
        return INSTANCE;
    }
}

File: src/main/kotlin/com/vandendaelen/nicephore/clipboard/impl/WindowsClipboardManagerImpl.kt
package com.vandendaelen.nicephore.clipboard.impl

import com.profesorfalken.jpowershell.PowerShell
import com.vandendaelen.nicephore.clipboard.ClipboardManager
import net.minecraft.Util
import java.io.File

class WindowsClipboardManagerImpl private constructor() : ClipboardManager {
    private val session: PowerShell? = if (Util.getPlatform() == Util.OS.WINDOWS) {
        PowerShell.openSession()
    } else {
        null
    }

    override fun clipboardImage(screenshot: File): Boolean {
        val command = """
            [Reflection.Assembly]::LoadWithPartialName('System.Drawing');
            [Reflection.Assembly]::LoadWithPartialName('System.Windows.Forms');
            
            ${'$'}filename = "${screenshot.absolutePath}";
            ${'$'}file = get-item(${'$'}filename);
            ${'$'}img = [System.Drawing.Image]::Fromfile(${'$'}file);
            [System.Windows.Forms.Clipboard]::SetImage(${'$'}img);
        """.trimIndent()

        return session?.executeCommand(command)?.isError?.not() ?: false
    }

    companion object {
        val instance: WindowsClipboardManagerImpl by lazy { WindowsClipboardManagerImpl() }
    }
}
